native/src/main/java/io/ballerina/lib/cdc/Listener.java

```diff
@@ -0,0 +1,235 @@
+import io.ballerina.lib.cdc.utils.Constants.NativeDataKeys;
+import io.ballerina.runtime.api.Environment;
+import io.ballerina.runtime.api.types.ObjectType;
+import io.ballerina.runtime.api.types.TypeTags;
+import io.ballerina.runtime.api.utils.StringUtils;
+import io.ballerina.runtime.api.utils.TypeUtils;
+import io.ballerina.runtime.api.values.BArray;
+import io.ballerina.runtime.api.values.BMap;
+import io.ballerina.runtime.api.values.BObject;
+import io.ballerina.runtime.api.values.BString;
+import io.debezium.engine.ChangeEvent;
+import io.debezium.engine.DebeziumEngine;
+import io.debezium.engine.format.Json;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static io.ballerina.lib.cdc.utils.Constants.ANN_CONFIG_TABLES;
+import static io.ballerina.lib.cdc.utils.Constants.ANN_NAME_EVENTS_FROM;
+import static io.ballerina.lib.cdc.utils.Constants.COLON;
+import static io.ballerina.lib.cdc.utils.ErrorUtils.createCdcError;
+import static io.ballerina.lib.cdc.utils.ErrorUtils.createError;
+import static io.ballerina.lib.cdc.utils.ModuleUtils.getPackageIdentifier;
+import static io.debezium.engine.DebeziumEngine.create;
+
+/**
+ * This class contains utility functions for the cdc:Listener object.
+ */
+public class Listener {
+
+    public static Object attach(BObject listener, BObject service) {
+        Object serviceMap = listener.getNativeData(NativeDataKeys.TABLE_TO_SERVICE_MAP);
+        Object serviceConfigAnn = getServiceConfigAnnotation(service);
+
+        ConcurrentHashMap<String, Service> updatedServiceMap = initializeServiceMap(serviceMap);
+
+        if (serviceConfigAnn == null) {
+            handleSingleServiceAttachment(serviceMap, updatedServiceMap, service);
+        } else {
+            updateServiceMap(serviceConfigAnn, service, updatedServiceMap);
+        }
+
+        listener.addNativeData(NativeDataKeys.TABLE_TO_SERVICE_MAP, updatedServiceMap);
+        return true;
+    }
+
+    private static Object getServiceConfigAnnotation(BObject service) {
+        return ((ObjectType) TypeUtils.getReferredType(TypeUtils.getType(service))).getAnnotation(
+                StringUtils.fromString(getPackageIdentifier() + COLON + ANN_NAME_EVENTS_FROM));
+    }
+
+    @SuppressWarnings("unchecked")
+    private static ConcurrentHashMap<String, Service> initializeServiceMap(Object serviceMap) {
+        if (serviceMap != null) {
+            return (ConcurrentHashMap<String, Service>) serviceMap;
+        }
+        return new ConcurrentHashMap<>();
+    }
+
+    private static void handleSingleServiceAttachment(Object serviceMap,
+                                                      ConcurrentHashMap<String, Service> updatedServiceMap,
+                                                      BObject service) {
```

Comment:
Usually we use the interface as the type unless we are going to use any implementation specific features. 
```suggestion
    private static void handleSingleServiceAttachment(Object serviceMap,
                                                      Map<String, Service> updatedServiceMap,
                                                      BObject service) {
```

Reply:
Fixed in [f5b00dc](https://github.com/ballerina-platform/module-ballerinax-cdc/pull/2/commits/f5b00dc08c0d2f8d76f3cbd121be6583f06526b1)